File: onyx-database/src/com/onyx/entity/SystemUser.java
package com.onyx.entity;

import com.onyx.persistence.IManagedEntity;
import com.onyx.persistence.annotations.Attribute;
import com.onyx.persistence.annotations.Entity;
import com.onyx.persistence.annotations.Identifier;

import javax.security.auth.Subject;
import java.security.Principal;

/**
 * Created by cosbor11 on 3/2/2015.
 *
 * User for the database
 */
@Entity
public class SystemUser extends AbstractSystemEntity implements IManagedEntity, Principal
{

    public SystemUser()
    {

    }

    @Identifier(loadFactor = 3)
    @Attribute
    protected String username;

    @Attribute(size = 255)
    protected String password;

    @Attribute
    private int roleOrdinal;

    private SystemUserRole role;

    public String getId() {
        return username;
    }

    public void setId(String id) {
        this.username = id;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    @SuppressWarnings("unused")
    public int getRoleOrdinal()
    {
        return roleOrdinal;
    }

    @SuppressWarnings("unused")
    public void setRoleOrdinal(int roleOrdinal)
    {
        this.roleOrdinal = roleOrdinal;
    }

    public SystemUserRole getRole()
    {
        role = SystemUserRole.values()[roleOrdinal];
        return role;
    }

    public void setRole(SystemUserRole role)
    {
        this.role = role;
        this.roleOrdinal = role.ordinal();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        SystemUser that = (SystemUser) o;

        return username != null ? username.equals(that.username) : that.username == null
                && (password != null ? password.equals(that.password) : that.password == null);

    }

    @Override
    public int hashCode() {
        int result = username != null ? username.hashCode() : 0;
        result = 31 * result + (password != null ? password.hashCode() : 0);
        return result;
    }

    @Override
    public String getName()
    {
        return username;
    }

    @Override
    public boolean implies(Subject subject)
    {
        return false;
    }
}

File: onyx-database/src/main/kotlin/com/onyx/entity/SystemUser.kt
package com.onyx.entity

import com.onyx.persistence.IManagedEntity
import com.onyx.persistence.annotations.Attribute
import com.onyx.persistence.annotations.Entity
import com.onyx.persistence.annotations.Identifier
import java.security.Principal
import javax.security.auth.Subject

/**
 * Created by Chris Osborn on 3/2/2015.
 *
 * User for the database.
 */
@Entity
data class SystemUser @JvmOverloads constructor(

    @Identifier
    var username: String? = null,

    @Attribute(size = 255)
    var password: String? = null,

    @Attribute
    var role: SystemUserRole? = null

) : AbstractSystemEntity(), IManagedEntity, Principal {

    @Attribute
    var firstName: String? = null

    @Attribute
    var lastName: String? = null

    @Suppress("UNUSED")
    @Attribute
    var phoneNumber: String?  = null

    @Suppress("UNUSED")
    @Attribute
    var emailAddress: String?  = null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val that = other as SystemUser?

        return if (username != null)
            username == that!!.username
        else
            that!!.username == null && if (password != null) password == that.password else that.password == null
    }

    // region Principal

    override fun getName(): String = username!!

    override fun implies(subject: Subject?): Boolean = false

    override fun hashCode(): Int {
        var result = username?.hashCode() ?: 0
        result = 31 * result + (password?.hashCode() ?: 0)
        result = 31 * result + (role?.hashCode() ?: 0)
        return result
    }

    // endregion
}